minimize git fetching time by caching a repo in config:source_cache

cosmicexplorer opened the pull request:
*Notes:*
- *This PR depends on #31102. This is the diff vs that change: https://github.com/cosmicexplorer/spack/compare/git-ref-repo-utils...cosmicexplorer:git-cache-unify-repos?expand=1, **so this PR is really only `+355 / -315` in size.***
- *The dependencies for this PR are being tracked in #31103.*

### Problem

This is a response to goal **3.A** described in @becker33's discussion at #20536:

> **3.A Mirror VCS software separately**
>
> This PR will establish the mirror subdirectories for each VCS system, and manage when we pull/fail for unfound git refs for each type of mirror. It will leave for future work the details for non-git VCS systems, while reserving their subdirectory names for that future use.

### Solution
1. Refactor `GitFetchStrategy` to create a new git *worktree* in each staging directory, after fetching the appropriate refs in a separate cached repo.
    - Reserve the cache subdirectory names for `svn` and `hg` for future use as in the prompt. This is done implicitly in the new `.persistent_cache_dir_for()` method in `FsCache`, which uses the fetcher's `.url_attr` as the first component of the relative cache path.
1. Implement the update logic described in #20536: tags and commits will never be re-fetched, while branches will be re-fetched if the remote repo is accessible.
1. Make several changes to `package.py` files to fix ambiguous `version()` directives (by adding `branch='master'`), as well as some other changes which have been isolated to the last commit for easier review.

trws commented on lib/spack/external/py2/typing.py:
```diff
@@ -1,28 +1,30 @@
 Annotated = None
 Any = None
-Callable = None
+Callable = defaultdict(lambda: object)
```
I like this workaround for being able to declare re-usable composite types.  Would you mind looking at a PR to finish this for the common types here?  It's a lot cleaner than what I'd been contemplating, and should help a lot with some of the nasty union types we have to deal with.

cosmicexplorer replied:
Yes absolutely! I think it would go best in #21720. I'm actually quite proud of this because I know the pex project just decided not to use `Union` or anything in its mypy migration because we (including me) hadn't figured it out yet.

becker33 commented on lib/spack/spack/test/conftest.py:
```diff
@@ -1234,3 +1239,18 @@ def mock_test_stage(mutable_config, tmpdir):
+
+
+@pytest.fixture(scope='function')
+def patch_git_ref(monkeypatch):
+
+    def from_ref(ref):
+        assert isinstance(ref, fs.GitRef), ref
+
+        def from_directive(*args, **kwargs):
+            return ref
+        # py2 complains unless you bind the .from_directive() classmethod to the class.
+        from_directive = from_directive.__get__(fs.GitRef, fs.GitRef.__class__)
+        monkeypatch.setattr(fs.GitRef, 'from_directive', from_directive)
+
+    return from_ref
```
This could use more explanation of what it's doing. I think I've parsed it, but it took more effort than is ideal for future maintainers.

I think it also might help if the fixture name were `patch_from_directive_for_git_ref` to make clearer what's going on.

becker33 replied:
Reading further and seeing how this is used, it might make more sense to use this as a factory -- it seems like it's always called with a new object, it could just take an argument for what sort of object to generate. Maybe something like `mock_git_ref_factory`, with a parameter "type" to choose among `'branch'`, `'tag'`, or `'commit'`.?

cosmicexplorer replied:
Yes, this is extremely difficult to follow. This would not be necessary at all if we ensured fetchers used in testing were always constructed with arguments sufficient to make `.from_directive()` pass. Now that I've gotten a handle on the unit tests passing, I think we can probably remove this fixture entirely and instead just construct fetch strategies with arguments that are sufficient to disambiguate the git ref they refer to. I want to continue raising a `FetcherConflict` in the `GitFetchStrategy` constructor, because erroring out earlier is always nicer.

I will try to just remove this fixture entirely. I think the result will be much easier to follow.

cosmicexplorer replied:
Ok, I reconsidered this, and the reason I would prefer not to do:
> Maybe something like `mock_git_ref_factory`, with a parameter "type" to choose among 'branch', 'tag', or 'commit'.?

is because to me that gets annoyingly ambiguous pretty fast, and the type `GitRef` was intended to be part of the solution to this general problem, by ensuring that none of our code (including tests) are relying on ambiguous behavior.

For example, I learned that if I set `init.defaultBranch` to `main` globally, all of our git tests fail, because they read from the user's global git config (*not* the one from the spack repo itself)! So I would really prefer to funnel test cases through the `GitRef` constructor to avoid this kind of issue.

However, I have renamed it to `patch_from_directive_for_git_ref` as you suggested above.

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -730,6 +734,113 @@ def __str__(self):
         return "[go] %s" % self.url
 
 
+class InvalidGitRef(ValueError):
```
We have a convention of putting error classes at the bottom of the file

cosmicexplorer replied:
What I was thinking: this exception type is not intended to ever go outside the current module, as it refers to an internal error and does not subclass `SpackError`.

However, it's super easy and makes sense to do both:
1. Move this to the bottom of the file.
1. Make it start with an underscore.

cosmicexplorer replied:
I have put this at the bottom of the file, and renamed it to start with an underscore!

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -730,6 +734,113 @@ def __str__(self):
+        # type: (str) -> GitRef
+        return cls('branch', ref)
+
+    # TODO: add test!
```
This looks vestigial

cosmicexplorer replied:
I think I would like to add an extremely tiny parameterized test for `GitRef.from_directive()` itself so bugs I might add in the future turn up more easily. I'll remove this comment in doing so.

cosmicexplorer replied:
Removed!

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -730,6 +734,113 @@ def __str__(self):
+    def refspec(self):
+        # type: () -> str
+        """Return a string which unambiguously identifies this ref's type."""
+        if self.eq_type(self.commit):
```
Couldn't this just as easily be `if self._ref_type == 'commit'`?

cosmicexplorer replied:
Yes! I have made this change.

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -730,6 +734,113 @@ def __str__(self):
+            # Commits are always specified via hash, and the hash unambiguously
+            # specifies it.
+            return self._ref
+        if self.eq_type(self.tag):
```
`if self._ref_type == 'tag'`

cosmicexplorer replied:
I believe your suggestion could actually make the code more difficult to read as well as significantly more prone to silent errors that could eat up a ridiculous amount of someone's time in the future, and I would like to avoid you, me, or anyone else having to spend that time in the future, hence the lengthy discussion below.

Someone else reading this code is already going to have to stop and understand the inherent complexity of the implementation, since we're fundamentally changing how we cache git repos. Here's what they would also have to think about if we applied your suggestion here:
```python
if self._ref_type == 'tag':
    ...
```

1. what is `self._ref_type`? It's a private variable, so could be anything.
1. what is `'tag'`? Is that a valid value for `self._ref_type` here? Did I spell the string correctly?
1. what does it mean for `self._ref_type` to be equal to `'tag'` here? Where is that described? What if that meaning changes later?

In contrast, the current code allows me to observe at a glance, without spooky action at a distance:
```python
if self.eq_type(self.tag):
    ...
```

1. `self.eq_type()` is checking what type of reference the current `self` is. This is clear from the name and from the fact that it is used everywhere else in this class.
1. `self.tag` will never, ever be misspelled, because mypy will error before runtime for a nonexistent attribute, and python will error at runtime. Both errors make it extremely clear what mistake was made.
1. Being equal to `self.tag` means we are the variant of `GitRef` which corresponds to a git tag. I know this because it is directly drawn from the name of the sole classmethod which is used in this file and in testing to *construct* tag refs.

The benefit we're seeing from this suggestion that would make up for the significantly increased risk of comparing private variables against string values is not clear to me. Just by *using* the method `self.eq_type()`, we get all of the benefits above, and we now can avoid having to flip back to the class's constructor every time we want to know what a variable is.

This is akin to the rationale for isolating this conditional logic into the `GitRef` class in the first place. We're not adding type checking, we are using an abstraction which lets us perform much more semantically-meaningful operations, such as calculating the argument to provide to git to update the specific refspec, without having to wonder whether I've correctly checked that `self` is supposed to be a `tag`.

Separately from anything else, I want to understand how we would respond to the case of a misspelled string, and how a user would be expected to diagnose that, and then I'd like to revisit the expected benefit of this suggestion to use `self._ref_type == 'tag'` instead of `self.eq_type(self.tag)`.

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -730,6 +734,113 @@ def __str__(self):
+            return self._ref
+        if self.eq_type(self.tag):
+            return 'refs/tags/{0}'.format(self._ref)
+        assert self.eq_type(self.branch), self
```
`self._ref_type == 'branch'`

cosmicexplorer replied:
Done!

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -730,6 +734,113 @@ def __str__(self):
+        self._ref_type = ref_type
+        self._ref = ref
+
+    def eq_type(self, fn):
```
I think this method can be easily replaced in every instance where it is used by a simple check against a string.

cosmicexplorer replied:
Yes!

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -768,173 +896,216 @@ def __init__(self, **kwargs):
+                tag=self.tag,
+                branch=self.branch,
+                version_name=self.version_name)
+        except InvalidGitRef as e:
+            raise FetcherConflict(
```
Is there a reason we change the type of the error here? Maybe I'll understand before I finish the review, but I'm not seeing it yet.

cosmicexplorer replied:
`GitRef.from_directive()` does a few things:
1. It defines the precise inputs that make up the concept of a symbolic "git reference".
2. It applies a set of precedence rules which define the `version()` matching logic.
3. It provides a specific error describing how its inputs failed to create an unambiguous "git reference", if applicable.

None of those things correspond to a `FetcherConflict`. It is only when we attempt to *parse the `GitRef` from the `version()` directive kwargs* that the "ambiguous git ref" error becomes a "fetcher failure" error. This is also why I made `InvalidGitRef` subclass `ValueError` and put it at the top of the file, since it's *not* supposed to ever be raised outside of this module -- we should be raising a subclass of `SpackError` for that.

cosmicexplorer replied:
Ok, after revising the thing a lot, the reason we now continue to change the type of the error is so that we can throw an `InvalidGitRef` exception without having to know the remote url. This means `GitRef.from_directive()` and other methods on `GitRef` can avoid having to pass in the url to construct their error strings.

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -768,173 +896,216 @@ def __init__(self, **kwargs):
     def cachable(self):
-        return self.cache_enabled and bool(self.commit or self.tag)
+        return (self.cache_enabled and
+                not self.ref.always_check_for_updates_but_accept_airgapped())
```
If we're going to use this in places not related to updating the cache, we need a better name for it.

Maybe something like `self.ref.is_mutable()` as an abstraction layer for `self.ref._ref_type == 'branch'` in case we change the behavior for tags or git adds a new type of mutable ref

cosmicexplorer replied:
`.is_mutable()` sounds great! We can keep the lengthy explanation in the `._fetch_ref()` method's docstring.

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -768,173 +896,216 @@ def __init__(self, **kwargs):
 
     def source_id(self):
-        return self.commit or self.tag
+        return self.ref.refspec()
```
This value is supposed to be a reproducible source reference, which we can't have in the case of a branch.

becker33 replied:
This may be `if not self.is_mutable(): self.ref.refspec()`

cosmicexplorer replied:
> This value is supposed to be a reproducible source reference, which we can't have in the case of a branch.

Ok, that makes sense! I'll do this verbatim.

*Separately, it was relatively somewhat difficult to follow the chain of logic that ends up consuming the package hash.*

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -768,173 +896,216 @@ def __init__(self, **kwargs):
+    def _dormant_branch_name(self, cache_path):
+        # type: (str) -> str
+        if self.branch == self._special_dormant_branch_name:
+            raise FetcherConflict(dedent("""\
```
We should probably have a deep-in-the-weeds configuration option that a user can specify to change the dormant branch name, and this error message should inform the user how to do this.

We should also put the word "spack" into the dormant branch name, so that conflicts are (even) less likely.

cosmicexplorer replied:
That sounds great! Will do.

cosmicexplorer replied:
Generating a UUID allows us to avoid this!

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -768,173 +896,216 @@ def __init__(self, **kwargs):
+            return False
+        # If this is the first time we've tried checking out this repo, we'll need to
+        # create this directory.
+        mkdirp(cache_path)
```
This will cause a problem if the directory already existed and was on a different branch. Is there a failure mode in which that is possible?

becker33 replied:
If the directory exists, I think we should check whether it is a git repo. If not, we should error out.

Then we should check whether it points to the same remote we want. If not, we should error out.

Then we can just checkout the dormant branch. This will eliminate the failure mode in which we end up on the wrong branch somehow, or a  user plays around looking at the different refs in the cache. I believe the only performance loss is in cases that we would error on the existing code, so I'm not too worried about the cost of these checks.

Is there something I'm missing in this?

cosmicexplorer replied:
> This will cause a problem if the directory already existed and was on a different branch. Is there a failure mode in which that is possible?

The `cache_path` is only ever set to the `__dormant` branch. If the `cache_path` is already initialized, then it is on the correct branch. We do not ever modify the checked-out `HEAD` of the cached git repo after creating it -- we will only *fetch* refs, which are then visible to every *worktree*. Worktrees are not at all considered nor do they matter in this method -- we strictly handle worktrees in the actual `.fetch()` method.

Worktrees will only ever be pointing to a single type of (immutable) ref, so the only time we ever need to worry about what's been in any type of git repo directory beforehand is in the case of *branches*. This is the reason I elected to use the extremely long name `.always_check_for_updates_but_accept_airgapped()`, because that conditional logic was the only case in which we have to consider the problem of mutable filesystem state. I think that the current name of that method isn't as helpful as it could be, but I *do* think that it is a good idea to make that conditional check into a verbose method name so people don't start relying on it elsewhere. "Whether the ref type is mutable" is really, really an implementation detail, imo, so I don't want to convert that judgement into a more generic predicate method.

cosmicexplorer replied:
I was able to work around this by generating a UUID!

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -768,173 +896,216 @@ def __init__(self, **kwargs):
+                trailing_args = [self.url, self.ref.updating_refspec()]
+            return self.git(
+                'fetch', verbosity_arg, submodules_arg,
+                *tuple(trailing_args))
```
Why convert to a tuple here?

cosmicexplorer replied:
If it works, great! I'll check that. I was under the impression python 2 (maybe 2.6?) strictly requires tuples to be used for positional splats.

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -768,173 +896,216 @@ def __init__(self, **kwargs):
+        # type: () -> None
+        """Fetch the ref specified by self.ref, with some rules for updating local refs.
+
+        For GitRef.tag(...) and GitRef.commit(...), we will never check for updates from
+        the remote. If we do not have the tag or commit SHA, we will pull it, and error
+        if we have no internet connection.
```
This should be written more clearly. For GitRef.tag and GitRef.commit, we will not check the remote unless the ref is absent locally ("check for updates" is ambiguous between checking when we have it and when we don't)

cosmicexplorer replied:
Ok, I will rewrite this to emphasize the contracts around specifically hitting the network as you describe!

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -1425,8 +1601,14 @@ class FsCache(object):
     def __init__(self, root):
         self.root = os.path.abspath(root)
 
+    # TODO: use this method to determine the cache path for url fetches! This would
+    # invalidate all cached artifacts -- would our users allow this?
```
I'm not entirely sure what this comment refers to.

cosmicexplorer replied:
A few things:
1. `FsCache.persistent_cache_dir_for()` is now actually defined in a way that supports persistent caching of *any* fetcher (as per the prompt to reserve `svn` and `hg` subdirectories for future use).
1. Despite that, we don't ever use this method for `UrlFetchStrategy` instances, because they use the `.fetcher()` method immediately below which requires that the fetcher be a `UrlFetchStrategy`. Instead of having two methods to do the same thing (find a subdirectory of the cache root), it would be really nice if all fetchers could use the *same* method.
1. We *could* right now implement `.fetcher()` using the new `.persistent_cache_dir_for()` method, *but* if we did that, it would start caching `UrlFetchStrategy` fetches into `var/spack/cache/_source-cache/archive/url/ab/ab..../` instead of where it currently does, which is `var/spack/cache/_source-cache/archive/ab/ab.../` (note the `/url/` component in the first and not the second).
    - It would be much cleaner and easier to follow if they all used the same method `.persistent_cache_dir_for()`, but since that change would modify the cache paths, that would effectively invalidate the entire url fetch cache for spack users. We could do some clever symlinking to avoid this, but I didn't want to introduce that complexity and just left it as having the two separate duplicate methods in `.fetcher()` and `.persistent_cache_dir_for()`.

cosmicexplorer replied:
I have reworded this to make it significantly more concise!

becker33 reviewed (changes requested):
A few direct change requests, and some clarification requests.

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -740,11 +745,412 @@ def __str__(self):
+        # (3) If no explicit git ref arguments are provided, assume that the version
+        # string itself is to be used as the branch name.
+        if (num_specified == 0) and version_name is not None:
+            return cls.branch(version_name)
```
I think this is good for now, but we should eventually replace this with something that infers the type automatically.

vsoch commented:
@cosmicexplorer if the package developer can now use this new syntax with tag, commit, etc., would it make sense to add this to the developer docs? Maybe somewhere in https://spack.readthedocs.io/en/latest/packaging_guide.html?

cosmicexplorer commented:
@vsoch yes absolutely!! Does that just mean changing a file in `src/docs/`?

cosmicexplorer commented:
To be clear, the syntax wasn't changed here, but ambiguous expressions such as `'False'` (which would evaluate to truthy) are now banned with a nice error message telling the user to use a `bool`. That can easily be added to the documentation.

vsoch commented:
@cosmicexplorer yes! I was thinking this file: https://github.com/spack/spack/blob/develop/lib/spack/docs/packaging_guide.rst, since that's where a developer would go to write a package, and these updates help with that.

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -740,11 +745,429 @@ def __str__(self):
+    def __call__(self, *args, **kwargs):
+        return self.git(*args, **kwargs)
+
+    @staticmethod
+    def _get_git_version(git):
+        # type: (Executable) -> Version
+        vstring = git('--version', output=str).lstrip('git version ')
+        return Version(vstring)
+
+    @classmethod
+    @memoized
+    def from_executable(cls, git):
+        # type: (Executable) -> ConfiguredGit
+        """Extract the version from a git executable and add defaults for compatibility.
+
+        This method being `@memoized` means that with the current implementation of
+        `Executable.__eq__`, any two `Executable` instances pointing to the same
+        executable file path will return a cached result from this method, which will
+        avoid traversing the filesystem.
+        """
+        version = cls._get_git_version(git)
+
+        # Disable advice for a quieter fetch
+        # https://github.com/git/git/blob/master/Documentation/RelNotes/1.7.2.txt
+        if version >= Version('1.7.2'):
+            git.add_default_arg('-c')
+            git.add_default_arg('advice.detachedHead=false')
+
+        # If the user asked for insecure fetching, make that work
+        # with git as well.
+        if not spack.config.get('config:verify_ssl'):
+            git.add_default_env('GIT_SSL_NO_VERIFY', 'true')
+
+        return cls(git, version)
```
This logic isn't working properly on macos.

```
$ spack -d fetch hwloc@master
<snip>
==> [2021-06-10-18:36:23.003847] '/usr/bin/git' '--version'
Traceback (most recent call last):
  File "/Users/becker33/spack/bin/spack", line 77, in <module>
    sys.exit(spack.main.main())
  File "/Users/becker33/spack/lib/spack/spack/main.py", line 771, in main
    return _invoke_command(command, parser, args, unknown)
  File "/Users/becker33/spack/lib/spack/spack/main.py", line 499, in _invoke_command
    return_val = command(parser, args)
  File "/Users/becker33/spack/lib/spack/spack/cmd/fetch.py", line 86, in fetch
    package.do_fetch()
  File "/Users/becker33/spack/lib/spack/spack/package.py", line 1372, in do_fetch
    self.stage.fetch(mirror_only, err_msg=err_msg)
  File "/Users/becker33/spack/lib/spack/spack/util/pattern.py", line 24, in __call__
    for item in self.container]
  File "/Users/becker33/spack/lib/spack/spack/util/pattern.py", line 24, in <listcomp>
    for item in self.container]
  File "/Users/becker33/spack/lib/spack/spack/stage.py", line 490, in fetch
    self.fetcher.fetch()
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 78, in wrapper
    return fun(self, *args, **kwargs)
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 1335, in fetch
    current_expanded_ref = self._ensure_local_ref()
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 1278, in _ensure_local_ref
    current_expanded_ref = self._maybe_expand_ref()
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 1259, in _maybe_expand_ref
    return self.canonical_git_repo.expand_commit_hash(self.ref)
  File "/Users/becker33/spack/lib/spack/llnl/util/lang.py", line 207, in _memoized_function
    func.cache[args] = func(*args)
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 1226, in canonical_git_repo
    configured_git = ConfiguredGit.from_executable(git_exe)
  File "/Users/becker33/spack/lib/spack/llnl/util/lang.py", line 207, in _memoized_function
    func.cache[args] = func(*args)
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 980, in from_executable
    version = cls._get_git_version(git)
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 967, in _get_git_version
    return Version(vstring)
  File "/Users/becker33/spack/lib/spack/spack/version.py", line 165, in __init__
    raise ValueError("Bad characters in version string: %s" % string)
ValueError: Bad characters in version string: 2.21.0 (Apple Git-122.2)

cosmicexplorer replied:
I added the method `_strip_apple_git_suffix` in a062080f75111be9693f264b2ad5f9f715ea1221 and this allows the command `spack fetch -d hwloc@master` (or indeed just `spack install hwloc@master`) to work correctly on macOS.

becker33 reviewed (changes requested):
I found two issues in testing.

One, the git output should be suppressed, so that it doesn't interfere with Spack output. It should be available on debug, but not by default.

```
$ spack fetch hwloc@master
Initialized empty Git repository in /g/g0/becker33/dev-spack/spack/var/spack/cache/git/https/github.com/open-mpi/hwloc.git/.git/
Switched to a new branch 'spack-internal-5ff53f6a-d45a-43d4-ba39-f30a72a2565f'
[spack-internal-5ff53f6a-d45a-43d4-ba39-f30a72a2565f (root-commit) 53db77a] 
fatal: Needed a single revision
==> Ref <on branch master> -> c51886bdf45b2a607e471c116484f8d5665ab507 was newly downloaded from https://github.com/open-mpi/hwloc.git.
Preparing worktree (detached HEAD c51886bd)
HEAD is now at c51886bd windows: warn if we find processor groups on a 32bits windows
$ spack fetch hwloc@master
==> Ref <on branch master> -> c51886bdf45b2a607e471c116484f8d5665ab507 did not have any updates from https://github.com/open-mpi/hwloc.git.
```

Two, it looks like versions from commits aren't being fetched from the persistent cache, they're coming from the normal cache
```
$ spack fetch clingo@spack
Initialized empty Git repository in /g/g0/becker33/dev-spack/spack/var/spack/cache/git/https/github.com/potassco/clingo.git/.git/
Switched to a new branch 'spack-internal-32f8e9b8-847e-4868-8489-464db10e8ff3'
[spack-internal-32f8e9b8-847e-4868-8489-464db10e8ff3 (root-commit) 774e47f] 
fatal: Needed a single revision
==> Ref <at commit 2a025667090d71b2c9dce60fe924feb6bde8f667> -> 2a025667090d71b2c9dce60fe924feb6bde8f667 was newly downloaded from https://github.com/potassco/clingo.git.
Preparing worktree (detached HEAD 2a02566)
HEAD is now at 2a02566 update changelog
$ spack fetch clingo@spack
==> Using cached archive: /g/g0/becker33/dev-spack/spack/var/spack/cache/_source-cache/git//potassco/clingo.git/2a025667090d71b2c9dce60fe924feb6bde8f667.tar.gz
==> Warning: Fetching from mirror without a checksum!
  This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!

cosmicexplorer commented:
@becker33:

> One, the git output should be suppressed, so that it doesn't interfere with Spack output. It should be available on debug, but not by default.

I believe I have fixed this completely now. `spack fetch hwloc@master` should now produce no output unless `config:debug:true` is on.

> Two, it looks like versions from commits aren't being fetched from the persistent cache, they're coming from the normal cache

That is intended behavior. The normal cache is checked first, then the persistent cache. If the normal cache entry is deleted, it falls back to the persistent cache. Here is the behavior on the current PR--note how we explicitly `rm` the normal `.gz` cache entry in the middle.

```bash
# cosmicexplorer@breeze: ~/tools/s1 5:15:50
; spack --verbose --stacktrace fetch clingo@spack
lib/spack/spack/fetch_strategy.py:1360 ==> Ref <at commit 2a025667090d71b2c9dce60fe924feb6bde8f667> -> 2a025667090d71b2c9dce60fe924feb6bde8f667 was newly downloaded from https://github.com/potassco/clingo.git.
  10.74s user 1.04s system 84% cpu 13.918 total
# cosmicexplorer@breeze: ~/tools/s1 5:17:43
; spack --verbose --stacktrace fetch clingo@spack
lib/spack/spack/fetch_strategy.py:657 ==> Using cached archive: /home/cosmicexplorer/tools/s1/var/spack/cache/_source-cache/git//potassco/clingo.git/2a025667090d71b2c9dce60fe924feb6bde8f667.tar.gz
lib/spack/spack/stage.py:551 ==> Warning: Fetching from mirror without a checksum!
  This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
  6.62s user 0.20s system 118% cpu 5.736 total
# cosmicexplorer@breeze: ~/tools/s1 5:18:46
; rm /home/cosmicexplorer/tools/s1/var/spack/cache/_source-cache/git//potassco/clingo.git/2a025667090d71b2c9dce60fe924feb6bde8f667.tar.gz

# cosmicexplorer@breeze: ~/tools/s1 5:19:05
; spack --verbose --stacktrace fetch clingo@spack
lib/spack/spack/fetch_strategy.py:1365 ==> Ref <at commit 2a025667090d71b2c9dce60fe924feb6bde8f667> -> 2a025667090d71b2c9dce60fe924feb6bde8f667 was already downloaded from https://github.com/potassco/clingo.git.
  7.42s user 0.33s system 112% cpu 6.862 total
```

> This logic isn't working properly on macos.

Looking into this now.

@vsoch:

> @cosmicexplorer yes! I was thinking this file: https://github.com/spack/spack/blob/develop/lib/spack/docs/packaging_guide.rst, since that's where a developer would go to write a package, and these updates help with that.

Will add some docs to this page!

cosmicexplorer commented:
@vsoch updated docs to reflect changes to `version()` directive parsing as well as the protocol for caching different types of git refs!

vsoch commented on lib/spack/docs/mirrors.rst:
```diff
@@ -266,6 +266,23 @@ is always searched first when attempting to retrieve files for an installation)
+Version Control Caching
+^^^^^^^^^^^^^^^^^^^^^^^
+
+In general, caching includes retrieved tarball archives and source control
```
I would suspect someone might land on these docs and not have context of being part of mirrors - I would add some detail here and open the section like "Spack uses a caching strategy to make retrieval of tarball archives and source repositories more efficient." (or something along those lines) so the first part of the section states exactly the context of the caching and why it's done.

cosmicexplorer replied:
Thank you! I have applied your suggestion directly, although I added a bit more clarification to the phrase "source repositories".

vsoch commented on lib/spack/docs/mirrors.rst:
```diff
@@ -266,6 +266,23 @@ is always searched first when attempting to retrieve files for an installation)
+
+#. A commit hash is immutable and can be cached indefinitely.
+
+#. A tag is *assumed to be immutable*, so when spack first resolves it to
```
Should we add something like "In practice this is not always the case, as tags can be re-created with a different history, however this case is less common." The "assumed to be" sort of implies that, but maybe we want to say it explicitly?

cosmicexplorer replied:
I have taken your suggestion and added a link to https://git-scm.com/docs/git-tag#_on_re_tagging which clarifies this really shouldn't be done!

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -95,6 +98,13 @@ def fetcher(cls):
 
 
+def _raise_from(new_exc, prev_exc):
+    # type: (BaseException, BaseException) -> None
```
Can you add a little docstring here to explain the purpose of this function?

cosmicexplorer replied:
Yes! Done.

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +802,460 @@ def __str__(self):
+    _ref = None                                 # type: str
+
+    @classmethod
+    def known_types(cls):
```
A question for my own curiository - what is the benefit to having this getter return cls._known_types vs. just interacting with a variable cls.known_types directly? Is it an effort to try to protect the actual variable from being touched (since it's "private" with an underscore?)

cosmicexplorer replied:
In general it is a way to perform "information hiding" by decoupling the accessor from the location of data, but in this case since the accessor isn't doing any validation or whatever, I agree it is unnecessary. I have also removed the unnecessary accessor for `_ref_type`! Thank you!

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +802,460 @@ def __str__(self):
+
+    def __init__(self, ref_type, ref):
+        # type: (str, str) -> None
+        if ref_type not in self.known_types():
```
Any reason that self.known_types is a method instead of a property (and then again why not just have this be a variable without the extra method or property?

cosmicexplorer replied:
Fixed as per above!

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +802,460 @@ def __str__(self):
+            raise TypeError('GitRef can only have the types {0}, but was given {1!r}'
+                            .format(self.known_types(), ref_type))
+        self._ref_type = ref_type
+        if not isinstance(ref, six.string_types):
```
Should we do both checks first before setting attributes on either ref_type or ref?

cosmicexplorer replied:
I find that logic tougher to follow in general, and more prone to getting subtly broken by someone else editing this code later, in comparison to a sequence of `[check->set, check->set]`. But I have actually in the meantime worked around the need to address this by making `ref_type` into a `ClassVar`!

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +802,460 @@ def __str__(self):
+        return '<{0}>'.format(self.repo_info_for_reference())
+
+    # Matches any 7-40 character hexadecimal string.
+    _commit_rx = re.compile(r'^[0-9a-f]{7,40}$', flags=re.IGNORECASE)
```
I added a commit version regex (slightly different but could be modified) so maybe when one of our PRs is merged we can have this be a global over class variable and share! The one I made is here (doesn't account for the shorter version) https://github.com/spack/spack/pull/24639/files#diff-8dbe8c0b0200a382d9363fe2f8a2133594db50baa72353853027cab92a014d47R49

cosmicexplorer replied:
So I decided to keep this separate even after rebasing over your changes, because while the result of a commit hash *from git output with `%H`* will always conform to `^[0-9af]{40}` (and I think we should care whether we are receiving that exact output), I think we want to be able to interpret commit hashes *from within a `package.py`* which may have upper or lower case, and not necessarily full 40-character hashes.

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +802,460 @@ def __str__(self):
+        if not isinstance(ref, six.string_types):
+            # The exception message from `re.match()` does not print the value when
+            # a non-string argument is provided, so we special-case that check here.
+            raise InvalidGitRef('git reference was not a string: {0}'.format(ref))
```
yeah and I'm thinking we could even organize all these git regexes / exceptions into a common area for other modules to use (at some future point).

cosmicexplorer replied:
I ended up consuming `InvalidGitRef` from within `version.py` after rebasing over your changes in `lookup_commit()` -- let me know if that makes sense to you!

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +802,460 @@ def __str__(self):
+        tag = kwargs.pop('tag', None)
+        branch = kwargs.pop('branch', None)
+        version_name = kwargs.pop('version_name', None)
+        # (2) Ensure no mutually exclusive kwargs are provided.
```
I like your numbered commenting! It helps me read/think through the steps logically.

cosmicexplorer replied:
:D :D :D :D :D!!!!

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +802,460 @@ def __str__(self):
+        num_specified = len(list(filter(None, [commit, tag, branch])))
+        # (3) If no explicit git ref arguments are provided, assume that the version
+        # string itself is to be used as the branch name.
+        if (num_specified == 0) and version_name is not None:
```
I think if this went through a Python brain linter some folks would say you can just do `if not num_specified` instead of explicitly checking for 0. But this works okay too. I don't think you need the extra parentheses.

cosmicexplorer replied:
I have been burned by python conditional syntax one too many times to feel comfortable about removing the parentheses, since `and` has surprisingly high precedence (or maybe it's `or`?)! The `== 0` is also a workaround for that, since I don't want someone reviewing this to have to think too hard to make sure it's validating things correctly, especially if this code gets edited later!

I am willing to change this (I believe you are correct that `not` has higher precedence than `and`), but I would really prefer to keep this the way it is as a form of documentation for the intent of the code.

vsoch replied:
Sure! I think I was just wondering why not do:

```python
if not num_specified and not version_name
```
We know num_specified is going to be a positive number or 0, and is a empty string version name allowed?

No worries if you don't want to change - what you have now is actually more readable. But I wanted to check that version_name isn't an empty string. Sometimes I often do:

```
version_name = version_name or None
version_name = version_name or ""
```
to make sure it's the exact empty-thing that I want :)

cosmicexplorer replied:
This was a toughie! Thank you for writing out all your reasoning!

In this case I explicitly wanted to avoid `version_name` being a blank string (which I'd *totally* failed to consider until you mentioned it!!), so I converted this to:
```python
        if (num_specified == 0) and version_name:
            return cls.Branch(version_name)
````

I've kept `(num_specified == 0)` because I'm e.g. comparing against `num_specified == 2` later and it feels more appropriate to maintain the same structure. I agree that it's overparenthesized but I think in this case it could be genuinely confusing to see the alternative:
```python
        if not num_specified and version_name:
            return cls.Branch(version_name)

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +802,460 @@ def __str__(self):
+    what operations are performed to *prepare* a git checkout for a spack `Stage`,
+    *after* fetching the configured ref.
+    """
+    submodules = None           # type: bool
```
These will always be shared among all classes instantiated at the same time? E.g., having them here vs. just set in init means that they will be.

cosmicexplorer replied:
So yes, but [this is the way to tell mypy that a class has an instance variable with this type](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#classes). By doing this, we have a neat little description of all the instance variables that a class will have (and their types), without having to scan the body of `__init__` (which may change). This ideally makes any later diffs that modify the class variables much easier to review, because regardless of how `__init__` changes, it will be clear at a glance which variables are set.

I can't easily find the documentation about this, but I recall the best practice for py2-compatible code is to do `= None` for instance variables in order to make use of this paradigm. When we require py3 we can just type `submodules: bool`.

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +802,460 @@ def __str__(self):
+
+        assert git('--version', output=str) == 'git version 2.31.1'
+    """
+    git = None                 # type: Executable
```
Same here.

cosmicexplorer replied:
See above!

vsoch reviewed (commented):
I only have time to do the top part (have to go for my run) but this is coming along nicely! Sorry for my slowness in reviewing.

cosmicexplorer commented:
Just rebased past @vsoch's awesome changes to calculate commit distance a couple months ago, will address their comments now!

vsoch commented:
welcome back @cosmicexplorer ! :tada:

cosmicexplorer commented:
Hm, I'm not able to reproduce the python3.7 failure on the `test_style` test on the `ubuntu18.04` dockerfile image, and running `spack style` by itself seems to pass just fine. How should I try to attack this next?

cosmicexplorer commented:
Ok, I'm finally able to reproduce the `test_style()` failure locally, so I'm going to investigate that now.

cosmicexplorer commented:
Ok, `test_style()` also appears to be failing on `develop`, which definitely isn't right. I haven't been able to completely build our ubuntu docker image locally yet so I'll try to do that again and see if I can correctly repro it now.

cosmicexplorer commented:
**FINALLY** figured out the weird test failure (fixed with 6cd268d1d2e004a041b8364e46eac216e249d6d3)!!! @becker33 @vsoch @trws @tgamblin: please review this again for mergeability if you get a moment!

vsoch commented on lib/spack/docs/mirrors.rst:
```diff
@@ -266,6 +266,29 @@ is always searched first when attempting to retrieve files for an installation)
+
+#. A branch is assumed to be **mutable**, which means spack will attempt to
+   re-fetch the branch every time the user attempts to fetch the package (or
+   perform another command like ``spack install`` which fetches the package). If
```
```suggestion
   perform another command like ``spack install`` that fetches the package). If
```

cosmicexplorer replied:
Yes! Thanks.

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +796,585 @@ def __str__(self):
+
+    def __init__(self, ref):
+        # type: (str) -> None
+        if self.ref_type not in self.known_types:
```
Does capitalization matter here? E.g., can we standardize so anything with the string will work (e.g., so here do lower() and then capitalize() ?

cosmicexplorer replied:
Oh well this part was just intended to be an especially obnoxious logical check -- when I say in the comment just below "it cannot arise from invalid user input" I mean it's not going to happen. It's more intended to be a reminder that GitRef should never be instantiated by itself, only through one of those three subclasses, known in advance.

It's possible this check could be removed, since it's really unimportant to the overall logic.

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +796,585 @@ def __str__(self):
+        if (num_specified == 0) and version_name is not None:
+            return cls.Branch(version_name)
+        # (4) If we are given a tag or a branch with a commit sha, note that by using
+        # a special constructor.
```
@cosmicexplorer your comments are *chef's kiss* :kissing_heart:  I wish there were more developers like yourself!

cosmicexplorer replied:
I don't want others to have to solve any of these problems after me so I make sure to write these things down!

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +796,585 @@ def __str__(self):
+            return cls.Commit(commit)
+        if tag is not None:
+            return cls.Tag(tag)
+        assert branch is not None, (commit, tag, branch, version_name)
```
Will this be weird for the user to see if the assertion fails? I guess they can just post a GitHub issue but maybe we could give them a more human friendly thing of how to fix it.

cosmicexplorer replied:
The assertion should not fail because we have done the prior check in `__init__()` that we discussed here: https://github.com/spack/spack/pull/22235#discussion_r810546522. Does that make sense?

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +796,585 @@ def __str__(self):
+    def refspec(self):
+        # type: () -> str
+        """Return a string which unambiguously identifies this ref to git."""
+        if self.ref_type == 'Commit':
```
We probably can't do this yet because spack has to support old janky python but it would be fun to use enum for this kind of thing :) https://docs.python.org/3/library/enum.html

cosmicexplorer replied:
Yes!!!! We use it a lot in pants and some in pip since they've moved to py3-only now too :)))

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +796,585 @@ def __str__(self):
+    def should_validate_matches_hash(self):
+        # type: () -> Optional[str]
+        """There is no need to verify that a git commit hash matches itself."""
+        return None
```
Won't just return (without None) still return None? is there an old Python version that doesn't support that?

cosmicexplorer replied:
Yes, but the return value is going to be read by another function, so I wanted to make it clear it wasn't going to be thrown away. You're right that that's already somewhat indicated by the type annotation `-> Optional[str]`.

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +796,585 @@ def __str__(self):
+            A valid commit hash case-insensitively matches the regular expression '{1}'.
+            7 hex characters is the size printed out by `git log --format='%h'`, while
+            40 hex characters is the size printed out by `git log --format='%H'`.
+            """.format(ref, self._commit_rx.pattern)))
```
We should check this isn't already provided (the regex) somewhere else in spack.

cosmicexplorer replied:
I tried to ensure we centralized the regex patterns in this PR! I'm pretty sure this is provided just in two places, and they're slightly different patterns:
```bash
# cosmicexplorer@terrestrial-gamma-ray-flash: ~/tools/s1 18:05:59 
; rg -i '0\-9a\-f'
lib/spack/spack/fetch_strategy.py
931:    _commit_rx = re.compile(r'^[0-9a-f]{7,40}$', flags=re.IGNORECASE)
1197:    _tag_mapping_rx = re.compile(r'^([0-9a-f]{40}) refs/tags/(.*)$')

share/spack/qa/flake8_formatter.py
51:            r'([\'"])[0-9a-fA-F]{32,}\1',  # long hex checksums

lib/spack/external/ruamel/yaml/resolver.py
213:    |[-+]?0x[0-9a-fA-F_]+
281:        |[-+]?0x[0-9a-fA-F_]+)$''', re.X),
288:        |[-+]?0x[0-9a-fA-F_]+

asdf/spack-src/build-aux/gitlog-to-changelog
169:          $line =~ /^([0-9a-fA-F]{40})$/
243:      $sha =~ /^[0-9a-fA-F]{40}$/

var/spack/repos/builtin/packages/binutils/update_symbol-2.26.patch
70:+ +[0-9]+: +[0-9a-f]+ +[0-9a-f]+ +FUNC +GLOBAL +DEFAULT +[0-9]+ +foo@VERS.1
72:+ +[0-9]+: +[0-9a-f]+ +[0-9a-f]+ +FUNC +GLOBAL +DEFAULT +[0-9]+ +foo@@VERS.2
```

where `_tag_mapping_rx` specifically searches for 40-character hexadecimal strings, and *not* `7-40` characters like `_commit_rx`. We could merge it with `_commit_rx`, but I personally appreciate having all that matching logic stored in the `_tag_mapping_rx` pattern.

Were you thinking of anywhere else that should be doing this logic differently?

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +796,585 @@ def __str__(self):
+            self.ref_type, self._ref, self._commit,
+        )
+
+GitRef.Tag = GitTag             # noqa: E305
```
Just a thought - and I'm sort of torn. I like this way because you can see the class added right after, but another idea is to group them together so it's more tidy. I guess I don't feel strongly either way but wanted to mention it in case someone else feels strongly.

cosmicexplorer replied:
Was also torn over the same thing!

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +796,585 @@ def __str__(self):
+        # type: (Any) -> GitFetchStageConfiguration
+        try:
+            submodules = cls._extract_bool(kwargs, 'submodules', False)
+            submodules_delete = kwargs.pop('submodules_delete', None)
```
This should default to None no?

cosmicexplorer replied:
Yes, and I believe that's what `kwargs.pop('submodules_delete', None)` is doing?

vsoch commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -792,11 +796,585 @@ def __str__(self):
+
+        assert git('--version', output=str) == 'git version 2.31.1'
+    """
+    git = None                 # type: Executable
```
Does this even get instantiated multiple times? E.g., if it's a class variable changing for one class will change the others too.

cosmicexplorer replied:
It begins as a class variable, but once `self.git` is assigned to within `__init__`, it becomes an instance variable! This `= None` method is the only way to annotate fields with mypy in python 2.

cosmicexplorer replied:
It's also possible to avoid the field type declaration, and simply rely on assigning `self.git` within `__init__()` to make its type known to mypy, so it's slightly redundant to have this field type declaration at the top. I like using it because it shows me what fields the object has at a glance.

vsoch reviewed (commented):
Ah started reviewing but I think you pushed changes - will submit what I have for now!

vsoch commented:
Sure thing, done! Also just noticed:

> Just rebased past @vsoch's awesome changes to calculate commit distance a couple months ago, will address their comments now!

Is the "their" referencing me? You can call me a she if you want, I mean I guess I don't do a good job of sharing my pronouns but I don't really feel like a "they" so maybe I should be better at that!

cosmicexplorer commented:
> You can call me a she if you want

This has been recorded in the database! Thank you for clarifying!

vsoch commented:
thank you!

- Vanessasaurus the honorable, avocado she-dinosaur.

vsoch commented:
Looks like you fixed it! 🎉

cosmicexplorer commented:
Oh my goodness! It turns out this PR would likely have made fetches *slower* than on `develop` until the change in b937ec7 was made, since I hadn't realized that `git fetch` also accepts the `--depth 1` argument! I only realized this when trying to fix the documentation for `get_full_repo` in `packaging_guide.rst`! This would have been incredibly annoying to fix later.

cosmicexplorer commented:
Ok, this is still failing a couple tests in `test/versions.py` since the `get_full_repo=False` fix. Working on fixing those now.

cosmicexplorer committed:
Update docs for new version control caching procedures

cosmicexplorer committed:
rewrite git fetch testing to use new git ref/repo helpers

cosmicexplorer committed:
allow for raw file paths in a git url

cosmicexplorer committed:
fix package definitions for new git caching

- add branch='master' to several places relying on it implicitly
- also ensure submodules='true' and other string arguments are not used
  - this is so that we can be sure that the only truthy values we receive are booleans, in general

cosmicexplorer committed:
modify GitFetchStrategy

cosmicexplorer committed:
add git ref/repo utils

cosmicexplorer committed:
adding types

alalazo commented:
Closing, since there has been no activity in a while. Feel free to reopen if you want to continue work on this.
